libyul/optimiser/VarNameCleaner.cpp

```diff
@@ -0,0 +1,134 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/VarNameCleaner.h>
+#include <libyul/AsmData.h>
+#include <algorithm>
+#include <cctype>
+#include <climits>
+#include <iterator>
+#include <string>
+
+using namespace yul;
+using namespace std;
+
+// ensure the following transforms
+// - input: a, a_1, a_1_2
+//	 output: a, a_1, a_2
+// - input: a, a_1, a_1_2, a_2
+//   output: a, a_1, a_2, a_3
+// - input: a_15, a_17
+//   output: a, a_2
+//
+// - input: abi_decode_256
+//   output: abi_decode
+
+void VarNameCleaner::run(Block& _ast)
+{
+	VarNameCleaner{}(_ast);
+}
+
+void VarNameCleaner::operator()(VariableDeclaration& _varDecl)
+{
+	for (TypedName& typedName: _varDecl.variables)
+	{
+		if (auto newName = makeCleanName(typedName.name.str()))
+		{
+			printf("VariableDeclaration: %s -> %s\n", typedName.name.str().c_str(), newName->c_str());
+			typedName.name = YulString{*newName};
+		}
+		else
+			printf("VariableDeclaration: %s\n", typedName.name.str().c_str());
+	}
+
+	ASTModifier::operator()(_varDecl);
+}
+
+void VarNameCleaner::operator()(Identifier& _identifier)
+{
+	if (auto newName = getCleanName(_identifier.name.str()))
+		_identifier.name = YulString{*newName};
+}
+
+VarNameCleaner::VarNameCleaner()
+{
+	printf("VarNameCleaner.ctor\n");
+}
+
+// make identifier clean, remember new name for reuse, keep them unique and simple
+boost::optional<string> VarNameCleaner::makeCleanName(string const& _name)
+{
+	if (auto newName = stripSuffix(_name))
+	{
+		printf("makeCleanName.1: %s -> %s (# %zu)\n", _name.c_str(), newName->c_str(),
+				m_usedNames.count(*newName));
+		if (!m_usedNames.count(*newName))
+		{
+			printf("makeCleanName.1: %s -> %s\n", _name.c_str(), newName->c_str());
+			m_usedNames[_name] = *newName;
+			m_usedNames[*newName] = *newName;
+			return newName;
+		}
+
+		// create new name with suffix (by finding a free identifier)
+		for (uintmax_t i = 1; i < numeric_limits<decltype(i)>::max(); ++i)
+		{
+			string newNameSuffixed = *newName + "_" + to_string(i);
+			if (!m_usedNames.count(newNameSuffixed))
+			{
+				printf("makeCleanName.2: %s -> %s\n", _name.c_str(), newNameSuffixed.c_str());
+				m_usedNames[_name] = newNameSuffixed;
+				m_usedNames[newNameSuffixed] = newNameSuffixed;
+				return newNameSuffixed;
+			}
+		}
+		yulAssert(false, "Exhausted by attempting to find an available suffix.");
+	}
+
+	// name isn't yet used and we're not going to rewrite it
+	printf("makeCleanName.3: %s\n", _name.c_str());
+	m_usedNames[_name] = _name;
+	return boost::none;
+}
+
+boost::optional<std::string> VarNameCleaner::stripSuffix(string const& _name)
+{
+	// valid suffixes are (_[0-9]+)+
+	auto isValidSuffix = [](string const& _name, size_t i) -> bool
+	{
+		return i + 1 < _name.size() && all_of(next(cbegin(_name), i),
```

Comment:
If you don't use a regex, this might be easier to read (and perhaps even more efficient) when rewritten as a loop over `i`.

Reply:
I take my previous comment back. I've decided to go `std::regex`, made it a 3-liner. :-)